core/resharding.go

```diff
@@ -47,38 +49,61 @@ func (ss *ShardingState) assignNewNodes(newNodeList []types.NodeID) {
 
 // cuckooResharding uses cuckoo rule to reshard X% of active committee(shards) into inactive committee(shards)
 func (ss *ShardingState) cuckooResharding(percent float64) {
-	ss.sortCommitteeBySize()
 	numActiveShards := ss.numShards / 2
-	kickedNodes := []types.NodeID{}
+	kickedNodes := []types.NodeInfo{}
 	for i := range ss.shardState {
 		if i >= numActiveShards {
 			break
 		}
-		Shuffle(ss.shardState[i].NodeList)
 		numKicked := int(percent * float64(len(ss.shardState[i].NodeList)))
-		tmp := ss.shardState[i].NodeList[:numKicked]
+		if numKicked == 0 {
+			numKicked++
+		}
+		length := len(ss.shardState[i].NodeList)
+		tmp := ss.shardState[i].NodeList[length-numKicked:]
 		kickedNodes = append(kickedNodes, tmp...)
-		ss.shardState[i].NodeList = ss.shardState[i].NodeList[numKicked:]
+		ss.shardState[i].NodeList = ss.shardState[i].NodeList[:length-numKicked]
 	}
 
 	Shuffle(kickedNodes)
+	numInactiveShards := ss.numShards - numActiveShards
 	for i, nid := range kickedNodes {
-		id := numActiveShards + i%(ss.numShards-numActiveShards)
+		id := numActiveShards + i%numInactiveShards
 		ss.shardState[id].NodeList = append(ss.shardState[id].NodeList, nid)
 	}
 }
 
 // UpdateShardState will first add new nodes into shards, then use cuckoo rule to reshard to get new shard state
-func (ss *ShardingState) UpdateShardState(newNodeList []types.NodeID, percent float64) {
+func (ss *ShardingState) assignLeaders() {
+	for i := 0; i < ss.numShards; i++ {
+		Shuffle(ss.shardState[i].NodeList)
+		for j := range ss.shardState[i].NodeList {
+			if j == 0 {
+				ss.shardState[i].NodeList[j].IsLeader = true
```

Comment:
need to check if coincident the older leader is selected again

Reply:
yep, it depends on we want that happen or not - now let's make it happen randomly - if we want to avoid that we can use Leader to check and then reshuffe